app/services/sso/basic.rb

```
@@ -21,5 +25,25 @@ def http_auth_set?
       request.authorization.present? && request.authorization =~ /\ABasic/
     end
 
+    def signo_auth
+      u, p = ActionController::HttpAuthentication::Basic.user_name_and_password(controller.request)
+      uri  = URI.parse("#{Setting['signo_url']}/login.json?username=#{URI.escape(u)}&password=#{URI.escape(p)}")
+      http = Net::HTTP.new(uri.host, uri.port)
+      if uri.scheme == 'https'
+        http.use_ssl     = true
+        http.verify_mode = OpenSSL::SSL::VERIFY_NONE
+      end
+      request  = Net::HTTP::Post.new(uri.request_uri)
+      response = http.request(request)
+      response.code == '200' ? User.find_by_login(u) : nil
+    rescue Timeout::Error, Errno::EINVAL, Errno::ECONNRESET, EOFError, Net::HTTPBadResponse,
```

Maybe catch all exceptions, if they have a common derivative?